### PR TITLE
[Segment Replication] Remove Doc Parsing for segment replication enabled replica shard during translog replay from recovery

### LINKED ISSUE #8848: [BUG][Segment Replication] Tests on searchQueryIT failing with seqNo [0] was processed twice in generation [2], with different data when segment replication enabled.
**Describe the bug**
Tests on searchQueryIT failing with seqNo [0] was processed twice in generation [2], with different data when segment replication enabled.

```
java.lang.AssertionError: seqNo [0] was processed twice in generation [2], with different data. prvOp [Index{id='1', seqNo=0, primaryTerm=1, version=1, autoGeneratedIdTimestamp=-1}], newOp [Index{id='1', seqNo=0, primaryTerm=1, version=1, autoGeneratedIdTimestamp=-1}]
	at __randomizedtesting.SeedInfo.seed([55718BEB7B387B03]:0)
	at org.opensearch.index.translog.TranslogWriter.assertNoSeqNumberConflict(TranslogWriter.java:338)
	at org.opensearch.index.translog.TranslogWriter.add(TranslogWriter.java:286)
	at org.opensearch.index.translog.Translog.add(Translog.java:571)
	at org.opensearch.index.translog.InternalTranslogManager.add(InternalTranslogManager.java:327)
	at org.opensearch.index.engine.NRTReplicationEngine.index(NRTReplicationEngine.java:208)
	at org.opensearch.index.shard.IndexShard.index(IndexShard.java:1082)
	at org.opensearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:1027)
	at org.opensearch.index.shard.IndexShard.applyTranslogOperation(IndexShard.java:2160)
	at org.opensearch.index.shard.IndexShard.applyTranslogOperation(IndexShard.java:2147)
	at org.opensearch.indices.recovery.RecoveryTarget.lambda$indexTranslogOperations$3(RecoveryTarget.java:294)
	at org.opensearch.action.ActionListener.completeWith(ActionListener.java:342)
	at org.opensearch.indices.recovery.RecoveryTarget.indexTranslogOperations(RecoveryTarget.java:269)
	at org.opensearch.indices.recovery.PeerRecoveryTargetService$TranslogOperationsRequestHandler.performTranslogOps(PeerRecoveryTargetService.java:490)
	at org.opensearch.indices.recovery.PeerRecoveryTargetService$TranslogOperationsRequestHandler.messageReceived(PeerRecoveryTargetService.java:441)
	at org.opensearch.indices.recovery.PeerRecoveryTargetService$TranslogOperationsRequestHandler.messageReceived(PeerRecoveryTargetService.java:424)
	at org.opensearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:106)
	at org.opensearch.transport.InboundHandler$RequestHandler.doRun(InboundHandler.java:453)
	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:908)
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
```

**To Reproduce**
`REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.search.query.SearchQueryIT" -Dtests.seed=55718BEB7B387B03 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=en -Dtests.timezone=Etc/UTC -Druntime.java=20`
 run with segment replication enabled.

**Expected behavior**

**Plugins**
Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by Rishikesh1159**
### Description
This PR removes removes Doc parsing step for segment replication enabled replica shard, which is called during translog replay in phase 2 of recovery.

There are two ways an operation can be written into replica shard's translog:
-> **First Way:** This is from [applyIndexOperationOnReplica()](https://github.com/opensearch-project/OpenSearch/blob/0750907e6276e36debf84b8fdd1328698c6a4cbe/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L938), where we avoid doc parsing and we set routing to null.
-> **Second way:** This is from [applyTranslogOperation()](https://github.com/opensearch-project/OpenSearch/blob/0750907e6276e36debf84b8fdd1328698c6a4cbe/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L2159), here we don't avoid doc parsing and parse actual routing value from input.

When phase 2 of recovery is started for a replica shard, it opens it's engine to write new translog operations (First way mentioned above )and then later we replay translog ops from primary shard (Second way mentioned above) during recovery. Sometimes a translog operation might be already written into replica's shard but it is still replayed from primary. Usually translog writer will be able to recognize that same operation is happening and does a noop if it is already present.
But sometimes in case of segrep replica shard, the translog writer is not able to recognize both of them as same operation because the routing value in one operation is null and routing value in another will be an actual parsed routing value. So we hit the following error #8848 sometimes.

To avoid this situation, this PR removes doc parsing even in second way on a segrep enabled replica shard. This way we completely remove doc parsing step for a segrep enabled replica shard.

### Related Issues
Resolves #8848 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Rishikesh1159** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/21464/
> * **CommitID:** [51a027b](https://github.com/opensearch-project/OpenSearch/commit/51a027b839d31fc89e04ecd54885756c48bfe143)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

#8887

File: server/src/main/java/org/opensearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -987,8 +972,27 @@ private Engine.IndexResult applyIndexOperation(
         long autoGeneratedTimeStamp,
         boolean isRetry,
         Engine.Operation.Origin origin,
-        SourceToParse sourceToParse
+        SourceToParse sourceToParse,
+        String id
     ) throws IOException {
+        if (indexSettings.isSegRepEnabled() && routingEntry().primary() == false && getReplicationEngine().isPresent()) {
```
Comment (Poojita-Raj) [inline server/src/main/java/org/opensearch/index/shard/IndexShard.java:978]: Nit: Add in a comment explaining why we are skipping doc parsing

Comment (mch2) [inline server/src/main/java/org/opensearch/index/shard/IndexShard.java:978]: nit - There is no need to validate the engine type here once checking the routing & index setting.

**Comment by mch2** [comment]
Thanks for catching this @Rishikesh1159.  While we are still working on these tests randomly using segrep, can we add a segrep version of the IT that caught this that always runs?

**Review by Poojita-Raj — APPROVED** [review]

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java [hunk 2]
```diff
@@ -1343,4 +1347,76 @@ public void testPrimaryReceivesDocsDuringReplicaRecovery() throws Exception {
         ensureGreen(INDEX_NAME);
         waitForSearchableDocs(2, nodes);
     }
+
+    public void testSimpleDFSQueryWithSegmentReplication() throws Exception {
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:1351]: nit - this test is less about the dfs query and more that we are indexing while the second node is recovering.  Can we rename to reflect this? Something like `testIndexWhileRecoveringReplica`.

**Review by mch2 — APPROVED** [review]
Thanks for finding this @Rishikesh1159 !  Only a nit so approved.

**Pull request merged**